File: PracticasPDOORuby/Practica4_Ruby/lib/Irrgarten/monster.rb
# frozen_string_literal: true

require_relative 'labyrinth_character'
require_relative 'dice'

module Irrgarten
  class Monster < LabyrinthCharacter
    # --- Constantes y Atributos de Lectura ---
    INITIAL_HEALTH = 5

    attr_reader :row, :col

    # ======================================================
    # MÉTODOS PÚBLICOS
    # ======================================================

    # Constructor
    # @param name Nombre del monstruo
    # @param intelligence Nivel de inteligencia del monstruo
    # @param strength Nivel de fuerza del monstruo
    # @return Objeto de la clase Monster
    def initialize(name, intelligence, strength)
      # Llama al constructor de la clase padre LabyrinthCharacter
      super(name, intelligence, strength, INITIAL_HEALTH)
    end

    # Calcula el ataque del monstruo (fuerza + dado de intensidad)
    # @return Valor del ataque del monstruo
    def attack
      Dice.intensity(@strength)
    end



    # ======================================================
    # P3 - Métodos públicos a implementar

    # Gestiona la defensa del monstruo ante un ataque.
    # Si el monstruo está vivo, calcula su defensa (con Dice.intensity).
    # Si el ataque supera la defensa, recibe daño (got_wounded).
    # @param received_attack
    # @return true si el monstruo muere a causa del ataque o ya estaba muerto, false en caso contrario
    def defend(received_attack)
      is_dead = dead # Diagrama 1.1: dead()

      # Diagrama opt: [!is_dead]
      unless is_dead # == if !is_dead
        defensive_energy = Dice.intensity(@intelligence) # Diagrama 1.2: intensity(intelligence)

        # Diagrama opt: [defensive_energy < received_attack]
        if defensive_energy < received_attack
          got_wounded # Diagrama 1.3: got_wounded()

          is_dead = dead # Diagrama 1.4: dead()
        else
          is_dead = false # No muere el monstruo
        end
      end

      is_dead # Diagrama 1.5: is_dead
    end

    # Representación en String del monstruo
    # @return String con la representación del monstruo
    def to_s
      "Monster " + super
    end
  end
end

File: PracticasPDOOJava/Practica4_Java/src/irrgarten/Monster.java
package irrgarten;

/**
 * Representa un monstruo en el laberinto.
 * <p>
 * Cada monstruo posee atributos básicos (nombre, inteligencia, fuerza y salud),
 * además de una posición (row y column) dentro del {@link labyrinth}
 * </p>
 * 
 * @autor Joaquín Cruz Lorenzo
 */
public class Monster extends LabyrinthCharacter {
    // ===============================
    // CONSTANTE DE CLASE
    // ===============================
    
    // Salud inicial por defecto para todos los monstruos.
    private static final int INITIAL_HEALTH = 5;
    
    // ELIMINO todos los atributos de instancia repetidos. 
    // Ya que los usamos en LabyrinthCharacter
    // ===============================
    // ATRIBUTOS DE INSTANCIA
    // ===============================
    // private String name;
    // private float intelligence, strength, health;
    // private int row = -1, col = -1; // No colocado inicialmente.
    
    // =================================
    // MÉTODOS PÚBLICOS
    // =================================
    
    /**
     * Constructor del Monstruo
     * Llama al constructor de la superclase con sus valores específicos.
     */
    public Monster(String name, float intelligence, float strength) {
        super(name, intelligence, strength, INITIAL_HEALTH);
    }
    
    /**
     * Calcula la potencia de ataque del monstruo.
     * <p> 
     * Se obtiene aplicando {@link Dice#intensity(float)} a su fuerza.
     * </p>
     * @return intensidad del ataque
     */
    public float attack() {
        return Dice.intensity(getStrength());
    }
    
     /**
     * Gestiona la defensa del monstruo ante un ataque.
     * <p>
     * Si el monstruo está vivo, calcula su defensa (con Dice.intensity).
     * Si el ataque supera a la defensa, recibe danio (gotWounded).
     * </p>
     * @param receivedAttack Ataque recibido
     * @return true si el monstruo muere a causa del ataque o ya estaba muerto,
     * false en caso contrario.
     */
    public boolean defend(float receivedAttack){ 
        boolean isDead = this.dead(); // Diagrama 1.1: dead()
        
        // Diagrama opt: [!isDead]
        if (!isDead) {
            float defensiveEnergy = Dice.intensity(getIntelligence()); // Diagrama 1.2: intensity(intelligence)
            
            // Diagrama opt: [defensiveEnergy < receivedAttack]
            if (defensiveEnergy < receivedAttack) {
                gotWounded(); // Diagrama 1.3: gotWounded()
                isDead = dead(); // Diagrama 1.4: dead()
            }
        }
        // Devuelve true si estaba muerto o si ha muerto ahora, 
        // false si ha sobrevivido al golpe
        return isDead;
    }
    
    @Override
    public String toString() { 
        return "Monster{" + super.toString() + "}";
    }
    
    // LISTA DE MÉTODOS SUPRIMIDOS YA QUE SE USAN AHORA EN LabyrinthCharacter
    // public boolean dead() { return health <= 0; }
    // public void setPos(int row, int col) {
    //    this.row = row;
    //    this.col = col;
    // }
   // private void gotWounded() { health--; }
}